File: streampark-flink/streampark-flink-packer/src/main/scala/org/apache/streampark/flink/packer/docker/DockerRetriever.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.flink.packer.docker

import org.apache.streampark.common.conf.{CommonConfig, InternalConfigHolder}
import org.apache.streampark.common.util.Utils

import com.github.dockerjava.api.DockerClient
import com.github.dockerjava.core.{DefaultDockerClientConfig, DockerClientConfig, HackDockerClient}
import com.github.dockerjava.httpclient5.ApacheDockerHttpClient

import java.net.URI
import java.time.Duration

object DockerRetriever {

  /**
   * docker config param from system properties, see in
   * https://github.com/docker-java/docker-java/blob/master/docs/getting_started.md#properties-docker-javaproperties
   * todo support custom docker configuration parameters in unified configurations in the future
   */
  lazy val dockerClientConf: DockerClientConfig = {
    DefaultDockerClientConfig.createDefaultConfigBuilder().build()
  }

  /**
   * docker http client builder, use ApacheDockerHttpClient by default todo support custom http
   * client configuration parameters in unified configurations in the future
   */
  private lazy val dockerHttpClientBuilder: ApacheDockerHttpClient.Builder =
    new ApacheDockerHttpClient.Builder()
      .dockerHost(dockerClientConf.getDockerHost)
      .sslConfig(dockerClientConf.getSSLConfig)
      .maxConnections(InternalConfigHolder.get(CommonConfig.DOCKER_MAX_CONNECTIONS))
      .connectionTimeout(
        Duration.ofSeconds(InternalConfigHolder.get(CommonConfig.DOCKER_CONNECTION_TIMEOUT_SEC)))
      .responseTimeout(Duration.ofSeconds(
        InternalConfigHolder.get(CommonConfig.DOCKER_RESPONSE_TIMEOUT_SEC)))

  /** get new DockerClient instance */
  def newDockerClient(): DockerClient = {
    setDockerHost()
    HackDockerClient.getInstance(dockerClientConf, dockerHttpClientBuilder.build())
  }

  /** set docker-host for kata */
  private def setDockerHost(): Unit = {
    val dockerHost: String = InternalConfigHolder.get(CommonConfig.DOCKER_HOST)
    if (Utils.isNotEmpty(dockerHost)) {
      val dockerHostUri: URI = new URI(dockerHost)
      dockerHttpClientBuilder.dockerHost(dockerHostUri)
    }
  }

}

File: streampark-flink/streampark-flink-packer/src/main/java/org/apache/streampark/flink/packer/docker/DockerRetriever.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.packer.docker;

import org.apache.streampark.common.configuration.Configuration;
import org.apache.streampark.common.configuration.GlobalConfiguration;
import org.apache.streampark.common.configuration.option.DockerOptions;
import org.apache.streampark.common.util.Utils;

import com.github.dockerjava.api.DockerClient;
import com.github.dockerjava.core.DefaultDockerClientConfig;
import com.github.dockerjava.core.DockerClientConfig;
import com.github.dockerjava.core.HackDockerClient;
import com.github.dockerjava.httpclient5.ApacheDockerHttpClient;

import java.net.URI;

/** Docker client factory. */
public final class DockerRetriever {

    public static final DockerClientConfig dockerClientConf =
        DefaultDockerClientConfig.createDefaultConfigBuilder().build();

    private DockerRetriever() {
    }

    /** Returns a new Docker client built from one consistent configuration snapshot. */
    public static DockerClient newDockerClient() {
        Configuration configuration = GlobalConfiguration.current();
        ApacheDockerHttpClient.Builder builder =
            new ApacheDockerHttpClient.Builder()
                .dockerHost(dockerClientConf.getDockerHost())
                .sslConfig(dockerClientConf.getSSLConfig())
                .maxConnections(configuration.get(DockerOptions.MAX_CONNECTIONS))
                .connectionTimeout(configuration.get(DockerOptions.CONNECTION_TIMEOUT))
                .responseTimeout(configuration.get(DockerOptions.RESPONSE_TIMEOUT));
        String dockerHost = configuration.get(DockerOptions.HOST);
        if (Utils.isNotEmpty(dockerHost)) {
            builder.dockerHost(URI.create(dockerHost));
        }
        return HackDockerClient.getInstance(dockerClientConf, builder.build());
    }
}
